packages/react-router/src/Matches.tsx

```diff
@@ -84,7 +69,35 @@ export interface RouteMatch<
   minPendingPromise?: ControlledPromise<void>
 }
 
-export type AnyRouteMatch = RouteMatch<any, any>
+export type MakeRouteMatch<
+  TRouteTree extends AnyRoute = RegisteredRouter['routeTree'],
+  TRouteId extends RouteIds<TRouteTree> = ParseRoute<TRouteTree>['id'],
+  TReturnIntersection extends boolean = false,
+  TTypes extends RouteById<TRouteTree, TRouteId>['types'] = RouteById<
+    TRouteTree,
+    TRouteId
+  >['types'],
+  TAllParams = TReturnIntersection extends false
+    ? TTypes['allParams']
+    : Expand<Partial<AllParams<TRouteTree>>>,
+  TFullSearchSchema = TReturnIntersection extends false
+    ? TTypes['fullSearchSchema']
+    : Expand<Partial<FullSearchSchema<TRouteTree>>>,
```

Comment:
Do we need to `Expand` here? I usually prefer to do it when the user actually needs the prettified type so it's done lazily